src/rules/noUnnecessaryBindRule.ts

```diff
@@ -0,0 +1,151 @@
+/**
+ * @license
+ * Copyright 2017 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as tsutils from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+export class Rule extends Lint.Rules.OptionallyTypedRule {
+    public static metadata: Lint.IRuleMetadata = {
+        description: "Prevents unnecessary and/or misleading scope bindings on functions.",
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        rationale: Lint.Utils.dedent`
+            \`function\` expressions that are immediately bound to \`this\` are equivalent to \`() =>\` arrow lambdas.
+            Additionally, there's no use in binding a scope to an arrow lambda, as it already has one.
+        `,
+        requiresTypeInfo: true,
+        ruleName: "no-unnecessary-bind",
+        type: "style",
+        typescriptOnly: false,
+    };
+
+    public static FAILURE_STRING_FUNCTION = "Don't bind \`this\` without arguments as a scope to a function. Use an arrow lambda instead.";
+
+    public static FAILURE_STRING_ARROW = "Don't bind scopes to arrow lambdas, as they already have a bound scope.";
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk);
+    }
+
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk, undefined, program.getTypeChecker());
+    }
+}
+
+function walk(context: Lint.WalkContext<void>, typeChecker?: ts.TypeChecker) {
+    function checkArrowFunction(node: ts.CallExpression, boundExpression: ts.Node): void {
+        if (node.arguments.length !== 1) {
+            return;
+        }
+
+        const replacement = Lint.Replacement.replaceNode(
+            node,
```

Comment:
If you only remove the text before and after the arrow function, you avoid overlapping fixes for nested failures.

Reply:
Intentionally adding overlapping fixes is the worst thing you can do. Go ahead and add a test for that case. You'll see that it destroys the code.

Reply:
Huh, interesting. Separately from this, shouldn't TSLint not merge those in? I suppose that's a very separate discussion. Will fix.

Reply:
Yeah, that's already discussed in #2556

A correct implementation could look like this: https://github.com/fimbullinter/wotan/blob/master/packages/wotan/src/fix.ts#L16